py-polars/polars/datatypes/classes.py

```diff
@@ -538,16 +538,31 @@ class Enum(DataType):
 
     categories: list[str]
 
-    def __init__(self, categories: list[str]):
+    def __init__(self, categories: list[str] | Iterator[str]):
         """
         A fixed set categorical encoding of a set of strings.
 
         Parameters
         ----------
         categories
-            Categories in the dataset.
+            Valid categories in the dataset.
 
         """
+        if isinstance(categories, Iterator):
+            categories = list(categories)
+
+        seen: set[str] = set()
+        for cat in categories:
+            if cat in seen:
+                raise ValueError(
+                    f"Enum categories must be unique; found duplicate {cat!r}"
+                )
+            if not isinstance(cat, str):
+                raise TypeError(
+                    f"Enum categories must be strings; found {cat!r} ({type(cat).__name__})"
+                )
+            seen.add(cat)
```

Comment:
I think we can simplify further if we build the list during the iteration, then we can accept an `Iterable[str]`:

```suggestion
        categories_list= []
        for cat in categories:
            if cat in categories_list:
                raise ValueError(
                    f"Enum categories must be unique; found duplicate {cat!r}"
                )
            if not isinstance(cat, str):
                raise TypeError(
                    f"Enum categories must be strings; found {cat!r} ({type(cat).__name__})"
                )
            categories_list.append(cat)
```

And then set `self.categories = categories_list`.

I think using a list is just as efficient as using a set, because the faster lookup is offset by the slower insert (not sure about this though).

Reply:
I'll test ;)

Reply:
Yup, confirmed - as I suspected, definitely shouldn't do list: `O(n)` vs `O(1)` lookup is not a favourable comparison, and the lookup costs utterly dominate  ;)